### ISSUE TITLE
May I Seek Your Insights on Whether Mask-Based Window Attention Effectively Lowers Computational Complexity?

### ISSUE THREAD

**Issue by Topdu** [QwenLM/Qwen2.5-VL#1049]
Hello,  

I hope this message finds you well. I have a question regarding the implementation of window-based attention, and I would greatly appreciate your insights.  

In principle, window attention is designed to reduce computational costs by limiting the attention computation to a local window of tokens. However, in some implementations, window attention is achieved using a masking approach. Specifically, the global $ QK^T $ is computed first, which incurs a computational cost of $ \text{Seq\_len}^2 \cdot \text{dim} $. Afterward, tokens outside the designated window are masked out during the softmax operation.  

On the other hand, if we were to compute attention only within the window (e.g., with a window size of 8), the computational cost would be $ 8^2 \cdot \text{dim} \cdot \text{Seq\_len} / 8 $, which is significantly lower. Based on this observation, I am inclined to think that implementing window attention through masking does not actually reduce the computational cost, as the global $ QK^T $ is still computed upfront.  

Could you kindly confirm whether my understanding is correct? If there are aspects I might be overlooking or alternative perspectives to consider, I would be very grateful for your guidance.  

Thank you so much for your time and expertise!

**Comment by hebangwen** [comment]
I also find that masked-attention is used in qwen2.5-vl instead of window attention. Thus I write this two ways of computing functions.

```py
def method1_mask(q, k, window_size):
    B, H, L, D = q.shape
    attn = torch.matmul(q, k.transpose(-2, -1))  # [B,H,L,L]

    mask = torch.full((L, L), float("-inf"), device=q.device)
    step = L // window_size
    for i in range(step):
        start = i * window_size
        end = min((i + 1) * window_size, L)
        mask[start:end, start:end] = 0

    attn += mask
    return F.softmax(attn, dim=-1)

def method2_window(q, k, window_size):
    B, H, L, D = q.shape
    num_windows = (L + window_size - 1) // window_size
    
    q_windows = q.unfold(2, window_size, window_size).permute(0, 1, 2, 4, 3)
    k_windows = k.unfold(2, window_size, window_size).permute(0, 1, 2, 4, 3)
    
    attn_blocks = torch.matmul(q_windows, k_windows.transpose(-1, -2))  # [B, H, W, N, N]
    attn_blocks = F.softmax(attn_blocks, dim=-1)
    
    attn = torch.zeros(B, H, L, L, device=q.device)
    for i in range(num_windows):
        start = i * window_size
        end = min(start + window_size, L)
        valid_size = end - start
        attn[:, :, start:end, start:end] = attn_blocks[:, :, i, :valid_size, :valid_size]
    
    return attn
```

It seems method2 is 1.7x faster that method1 in my benchmark due to less computation.

**Comment by JJJYmmm** [comment]
@Topdu @hebangwen Actually, the window attention implemented in qwen2.5vl NaViT is not mask-based. It follows the style of flash attention e.g. `cu_seqlens`. A image would be split into serval window images.

https://github.com/huggingface/transformers/blob/794fde7b1c3d041519fc28ea3e1461b0cfcad4e7/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L484C13-L491C54

**Comment by hebangwen** [comment]
> [@Topdu](https://github.com/Topdu) [@hebangwen](https://github.com/hebangwen) Actually, the window attention implemented in qwen2.5vl NaViT is not mask-based. It follows the style of flash attention e.g. `cu_seqlens`. An image would be split into serval window images.
> 
> https://github.com/huggingface/transformers/blob/794fde7b1c3d041519fc28ea3e1461b0cfcad4e7/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L484C13-L491C54

Thank you for your sharing, I see that cu_seqlens is called in flash_attention in default settings. In the case of sdpa and eager, the [implementation](https://github.com/huggingface/transformers/blob/794fde7b1c3d041519fc28ea3e1461b0cfcad4e7/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L300) is indeed mask-based. However, under flash_attention, the function called  `flash_attn_varlen_func`, I would like to ask if it will implement window-attention?

**Comment by JJJYmmm** [comment]
@hebangwen  Thank you for your clearify! `flash_attn_varlen_func` can implement the real window-attn because the `cu_seqlens` is define as `The cumulative sequence lengths of the sequences in the batch `

https://github.com/Dao-AILab/flash-attention/blob/aa04de66e22fb1810eeede8ba736ccd895f16274/flash_attn/flash_attn_triton_amd/fwd_ref.py#L181-L190

I will try to implement the real window-attn in the eager/sdpa version!

**Comment by JJJYmmm** [comment]
@hebangwen  @Topdu Hello! I’ve implemented the changes and conducted a test, with the results provided below.  
The measurements were taken while performing a grounding task using an 1000x2350 image. The process was repeated 20 times for consistency.

```bash
**flash attn:**
Inference time: 1.695 seconds
Peak GPU memory allocated: 17075.66 MB

**eager:**
Inference time: 4.170 seconds
Peak GPU memory allocated: 38803.97 MB
**eager(modified):**
Inference time: 3.284 seconds
Peak GPU memory allocated: 36530.72 MB

**sdpa:**
Inference time: 5.704 seconds
Peak GPU memory allocated: 36950.72 MB
**sdpa(modified):**
Inference time: 3.204 seconds
Peak GPU memory allocated: 36530.72 MB
```

The modified attention mechanisms exhibit lower memory usage and improved speed!!!

Thank you all for highlighting this! We will update it as soon as possible.

## modified eager/sdpa attention
```python
class Qwen2_5_VLVisionAttention(nn.Module):
    def __init__(self, dim: int, num_heads: int = 16) -> None:
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = dim // num_heads
        self.qkv = nn.Linear(dim, dim * 3, bias=True)
        self.proj = nn.Linear(dim, dim)

    def forward(
        self,
        hidden_states: torch.Tensor,
        cu_seqlens: torch.Tensor,
        rotary_pos_emb: Optional[torch.Tensor] = None,
        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
    ) -> torch.Tensor:
        seq_length = hidden_states.shape[0]
        q, k, v = self.qkv(hidden_states).reshape(seq_length, 3, self.num_heads, -1).permute(1, 0, 2, 3).unbind(0)
        if position_embeddings is None:
            logger.warning_once(
                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
                "removed and `position_embeddings` will be mandatory."
            )
            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
            cos = emb.cos()
            sin = emb.sin()
        else:
            cos, sin = position_embeddings
        q, k = apply_rotary_pos_emb_vision(q, k, cos, sin)

        # attention_mask = torch.full(
        #     [1, seq_length, seq_length], torch.finfo(q.dtype).min, device=q.device, dtype=q.dtype
        # )
        # for i in range(1, len(cu_seqlens)):
        #     attention_mask[..., cu_seqlens[i - 1] : cu_seqlens[i], cu_seqlens[i - 1] : cu_seqlens[i]] = 0

        # q = q.transpose(0, 1)
        # k = k.transpose(0, 1)
        # v = v.transpose(0, 1)
        # attn_weights = torch.matmul(q, k.transpose(1, 2)) / math.sqrt(self.head_dim)
        # attn_weights = attn_weights + attention_mask
        # attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(q.dtype)
        # attn_output = torch.matmul(attn_weights, v)
        # attn_output = attn_output.transpose(0, 1)
        # attn_output = attn_output.reshape(seq_length, -1)
        # attn_output = self.proj(attn_output)

        q = q.transpose(0, 1)
        k = k.transpose(0, 1)
        v = v.transpose(0, 1)

        attn_output = []
        for i in range(len(cu_seqlens) - 1):
            start = cu_seqlens[i]
            end = cu_seqlens[i + 1]

            q_window = q[:, start:end, :]
            k_window = k[:, start:end, :]
            v_window = v[:, start:end, :]

            attn_weights = torch.matmul(q_window, k_window.transpose(1, 2)) / math.sqrt(self.head_dim)
            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(q.dtype)
            attn_output_window = torch.matmul(attn_weights, v_window)
            attn_output.append(attn_output_window)

        attn_output = torch.cat(attn_output, dim=1).transpose(0, 1)
        attn_output = attn_output.reshape(seq_length, -1)

        attn_output = self.proj(attn_output)

        return attn_output


class Qwen2_5_VLVisionSdpaAttention(nn.Module):
    def __init__(self, dim: int, num_heads: int = 16) -> None:
        super().__init__()
        self.num_heads = num_heads
        self.qkv = nn.Linear(dim, dim * 3, bias=True)
        self.proj = nn.Linear(dim, dim)

    def forward(
        self,
        hidden_states: torch.Tensor,
        cu_seqlens: torch.Tensor,
        rotary_pos_emb: Optional[torch.Tensor] = None,
        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
    ) -> torch.Tensor:
        seq_length = hidden_states.shape[0]
        q, k, v = self.qkv(hidden_states).reshape(seq_length, 3, self.num_heads, -1).permute(1, 0, 2, 3).unbind(0)
        if position_embeddings is None:
            logger.warning_once(
                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
                "removed and `position_embeddings` will be mandatory."
            )
            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
            cos = emb.cos()
            sin = emb.sin()
        else:
            cos, sin = position_embeddings
        q, k = apply_rotary_pos_emb_vision(q, k, cos, sin)

        # attention_mask = torch.zeros([1, seq_length, seq_length], device=q.device, dtype=torch.bool)
        # for i in range(1, len(cu_seqlens)):
        #     attention_mask[..., cu_seqlens[i - 1] : cu_seqlens[i], cu_seqlens[i - 1] : cu_seqlens[i]] = True
        # q = q.transpose(0, 1)
        # k = k.transpose(0, 1)
        # v = v.transpose(0, 1)
        # attn_output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)
        # attn_output = attn_output.transpose(0, 1)
        # attn_output = attn_output.reshape(seq_length, -1)
        # attn_output = self.proj(attn_output)

        q = q.transpose(0, 1)
        k = k.transpose(0, 1)
        v = v.transpose(0, 1)

        attn_output = []
        for i in range(len(cu_seqlens) - 1):
            start = cu_seqlens[i]
            end = cu_seqlens[i + 1]

            q_window = q[:, start:end, :]
            k_window = k[:, start:end, :]
            v_window = v[:, start:end, :]

            attn_output_window = F.scaled_dot_product_attention(q_window, k_window, v_window, dropout_p=0.0)
            attn_output.append(attn_output_window)

        attn_output = torch.cat(attn_output, dim=1).transpose(0, 1)
        attn_output = attn_output.reshape(seq_length, -1)
        attn_output = self.proj(attn_output)

        return attn_output
```
